### ISSUE TITLE
[Feature Request] [Pull Request?] Declare Request._requestId or Request._interceptionId in the API

### ISSUE THREAD

**Issue by GregRos** [puppeteer/puppeteer#1708]
When you intercept or inspect a request via the DevTools protocol, you use a special `InterceptionId` or `RequestId`. Can you expose these properties via the `Request` object? I can already see that they are kept tracked of in the `_interceptionId` and `_requestId` fields, respectively, but they are undocumented and meant to be private.

Since this is a standard feature of the DevTools protocol, it seems to make sense to expose them. If it's okay by you, I will submit a pull request that does this myself.

Use cases:
1. Allows using `puppeteer` and a lower-level DevTools client together for things `puppeteer` does not support.
2. The properties are useful in and of themselves. For example, from my research, `_interceptionId` can be used to keep track of a series of redirects.

**Comment by aslushnikov** [comment]
@GregRos The API you request should be a part of a bigger story about `pptr<->other protocol clients` interop. We have a tracking bug for it - see #31, but there's no good plan yet. I'd love to have a general strategy first before exposing any specific bits of API.

Talking specifically about `interceptionId`, it belongs to the experimental part of the devtools protocol and will be removed in future: we're working on replacing it with `requestId` so that browser-side network stack could be matched against renderer-side counterparts.

> The properties are useful in and of themselves. For example, from my research, _interceptionId can be used to keep track of a series of redirects.

The same should be possible with `requestId` and over the pptr api.

> Allows using puppeteer and a lower-level DevTools client together for things puppeteer does not support.

Can you please share some of your usecases? Oftentimes, the required functionality should be exposed in pptr right away.

**Comment by GregRos** [comment]
Hmm, are you saying you can keep track of a series of redirects using the pptr API right now? How can I do that?

**Comment by aslushnikov** [comment]
@GregRos you can match requests by URLs and response status codes using [`'request'`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-request) and [`'response'`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-response) events; will this be enough?

**Comment by GregRos** [comment]
That's not nearly as good as using a built-in ID. It has a lot of room for error in principle and requires quite a bit more code besides, code that's pretty error-prone by itself.

**Comment by aslushnikov** [comment]
@GregRos will you be happy with the `request.redirectChain()` api? For example

```js
const response = await page.goto('http://google.com');
const requests = response.request().redirectChain();
for (const request of requests)
  console.log(request.url());
```

**Comment by GregRos** [comment]
Oh, sorry for not responding. It seems good but I'm not sure it covers my specific use-case. Here is what I was doing.

I want every request the browser makes to pass through my own code, so I can replace it with another resource, modify the response, etc. So what I did is set up request interception and added a `request` handler. When it was called, it would get all the info about the request (headers, method, body, etc) and pass it to a `ProxyEngine` that actually sent the request to the server. I get the response and pass it to the browser via the interception API.

`ProxyEngine` doesn't follow redirects on purpose, so sometimes it would return a redirect to the browser. In that case, the browser would then make a second request to the resource at the redirected location. This request would also go through the `request` handler. The problem is that, in the handler, I need to be able to tell if a request is original or if it's part of a chain of redirects from another original request.

To do this, I used the `_interceptionId`. So I need to follow a chain of redirects "as it's happening." Although the explanation above is a bit complicated, I think this feature makes sense as part of `puppeteer`. 

---

If your API can work even on the `Request` object passed in the handler, and the parent requests always maintain reference identity with the objects that were sent to the handler last time, that's great! It's also possible to lose reference identity but expose some unique ID (but in that case, why not use the same ID as in the protocol?)

**Comment by aslushnikov** [comment]
@GregRos thank you for taking time and explaining me your usecase.

> If your API can work even on the Request object passed in the handler, and the parent requests always maintain reference identity with the objects that were sent to the handler last time, that's great!

@GregRos this would be the case. Closing this in favor of #1816.

**Comment by GregRos** [comment]
@aslushnikov That's great! I'm looking forward to contributing something myself.

**Comment by GregRos** [comment]
Well, just a follow up:

I abandoned the notion doing this, and indeed of using request interception in Chromium, because it doesn't always work... some requests aren't routed through. I verified this by connecting Chromium to an HTTP proxy and saw it making requests by itself, without telling me. I guess neither you nor the documentation were kidding when you said it was experimental. 

In the end, I set up my own HTTP proxy, which Chromium is unlikely to circumvent.

**Comment by aslushnikov** [comment]
>  some requests aren't routed through.

@GregRos can you please specify what requests are not intercepted? We'll probably want to fix this

**Comment by GregRos** [comment]
Oh, sure. I've seen some more issues earlier, but I don't remember them very clearly.

This issue I can characterize pretty well.
If after you intercept the request and replace it with a raw response that has a 3xx redirect, the next request made for that redirect will not be intercepted. However, the request after that *will* be intercepted. I tried lots of different 3xx codes but they all had the same problem.

There are also some other requests that aren't intercepted, but I haven't been able to pin them down.

Another thing I noticed. Not sure if it's a bug or just weird:

The handler for the `request` event will not normally receive the fragment portion of the URL the browser wants (I say this because PhantomJS does allow you to receive the fragment). This is unfortunate. However, strangely enough, if you use `continue` to redirect to a page that returns a 3xx, and in the Location header you put a URL with a fragment, the next intercepted request *will* have a fragment in the URL. However, doing this will erase the original fragment of the URL.
